### PR TITLE
Feat: tw analytics

### PR DESCRIPTION
**Pull request by ahuininga-orisha**
Features: 
- Added setting to send productview/search and purchase requests to analytics api. With this setting enabled an separate javascript tracking script is not needed anymore for personal merchandising to work.

### CODE REVIEW

**Comment by ahuininga-orisha** [comment]
@jansentjeu Modified it so that the tw request is handled by php. And moved the search personalisation to an diffrent pull request

File: Block/Catalog/PersonalMerchandisingAnalytics.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+namespace Tweakwise\Magento2Tweakwise\Block\Catalog;
+
+use Magento\Framework\View\Element\Template;
+use Tweakwise\Magento2Tweakwise\Model\Config;
+use Tweakwise\Magento2TweakwiseExport\Model\Helper;
+use Magento\Store\Model\StoreManagerInterface;
+
+class PersonalMerchandisingAnalytics extends Template
+{
+    public function __construct(
```
Comment (jansentjeu) [inline Block/Catalog/PersonalMerchandisingAnalytics.php]: Add phpdoc to all functions

File: Block/Catalog/PersonalMerchandisingAnalytics.php
Comment (jansentjeu) [inline Block/Catalog/PersonalMerchandisingAnalytics.php:1]: Use a `ViewModel` if the class is only used to serve data. See https://m.academy/articles/magento-block-class-vs-view-model/

File: Model/Client.php [hunk 2]
```diff
@@ -114,10 +110,34 @@ protected function getClient(): HttpClient
...
     protected function createHttpRequest(Request $tweakwiseRequest): HttpRequest
+    {
+        if ($tweakwiseRequest->isPostRequest()) {
+            return $this->createPostRequest($tweakwiseRequest);
+        } else {
+            return $this->createGetRequest($tweakwiseRequest);
+        }
+    }
+
+    protected function createPostRequest(Request $tweakwiseRequest): HttpRequest
```
Comment (jansentjeu) [inline Model/Client.php:126]: Add phpdoc to functions

File: Model/Client/Request.php [hunk 3]
```diff
@@ -326,4 +326,15 @@ public function setProfileKey(string $profileKey)
         $this->setParameter('tn_profilekey', $profileKey);
         return $this;
     }
+
+    public function setParameterArray(string $parameter, array $value)
```
Comment (jansentjeu) [inline Model/Client/Request.php:333]: Add phpdoc

File: etc/frontend/di.xml [hunk 4]
```diff
@@ -21,6 +21,14 @@
             </argument>
         </arguments>
     </type>
+    <virtualType name="Tweakwise\Magento2Tweakwise\Block\SearchResult\ListProduct"
```
Comment (jansentjeu) [inline etc/frontend/di.xml]: Personal merchandising code

File: view/frontend/web/js/product-page.js
Comment (jansentjeu) [inline view/frontend/web/js/product-page.js:1]: product-page.js and search-page.js look very similar. Can you create 1 generic file for this that you can use for both cases?

**Review by jansentjeu — CHANGES_REQUESTED** [review]

File: Block/Catalog/PersonalMerchandisingAnalytics.php (hunk 1 above)
Comment (ahuininga-orisha) [reply Block/Catalog/PersonalMerchandisingAnalytics.php]: Changes implemented

File: Block/Catalog/PersonalMerchandisingAnalytics.php
Comment (ahuininga-orisha) [reply Block/Catalog/PersonalMerchandisingAnalytics.php:1]: Changes implemented

File: Model/Client/Request.php (hunk 3 above)
Comment (ahuininga-orisha) [reply Model/Client/Request.php:333]: Changes implemented

File: Model/Client.php (hunk 2 above)
Comment (ahuininga-orisha) [reply Model/Client.php:126]: Changes implemented

File: view/frontend/web/js/product-page.js
Comment (ahuininga-orisha) [reply view/frontend/web/js/product-page.js:1]: Changes implemented

File: etc/frontend/di.xml (hunk 4 above)
Comment (ahuininga-orisha) [reply etc/frontend/di.xml]: Changes implemented

File: Controller/Ajax/Analytics.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        private JsonFactory $resultJsonFactory,
+        private Client $client,
+        private PersonalMerchandisingConfig $config,
+        private RequestFactory $requestFactory
+    ) {
+        parent::__construct($context);
+    }
+
+    /**
+     * @return \Magento\Framework\App\ResponseInterface|\Magento\Framework\Controller\Result\Json|\Magento\Framework\Controller\ResultInterface
```
Comment (jansentjeu) [inline Controller/Ajax/Analytics.php]: Add classes as `use` statement

File: Model/Client.php [hunk 6]
```diff
@@ -114,10 +110,42 @@ protected function getClient(): HttpClient
...
+    {
+        $path = $tweakwiseRequest->getPath();
+        $headers = [];
+
+        $headers['Content-Type'] = 'application/json';
+        $headers['Instance-Key'] = $this->config->getGeneralAuthenticationKey();
+        $body = json_encode($tweakwiseRequest->getParameters());
+        //post request are used for the analytics api
+        $uri = new Uri($tweakwiseRequest->getApiUrl() . '/' . $path);
```
Comment (jansentjeu) [inline Model/Client.php]: Can we use the URL builder here?

File: view/frontend/templates/product/page.phtml
Comment (jansentjeu) [inline view/frontend/templates/product/page.phtml:1]: We can make it more generic. If we add an argument like `analyticsType` in the layout, than we can check this in the phtml to add the product key value when type is `product` and search query value when type is `search`.
Example:

```
<?php

/** @var \Tweakwise\Magento2Tweakwise\Block\Catalog\PersonalMerchandisingAnalytics $block */
/** @var  Escaper $escaper */

/** @var \Tweakwise\Magento2Tweakwise\ViewModel\PersonalMerchandisingAnalytics $viewModel */
$viewModel = $block->getData('viewModel');
$analyticsType = $block->getData('analyticsType');
$value = $analyticsType === 'product' ? $viewModel->getProductKey() : $viewModel->getSearchQuery();
?>

<script type="text/x-magento-init">
    {
        "*": {
            "Tweakwise_Magento2Tweakwise/js/analytics": {
              "value": "<?= $escaper->escapeHtmlAttr($value); ?>",
              "type": "<?= $analyticsType; ?>"
        }
    }
}
</script>
<script src="<?= $escaper->escapeHtmlAttr($block->getViewFileUrl('Tweakwise_Magento2Tweakwise::js/analytics.js')); ?>"></script>
```

**Review by jansentjeu — CHANGES_REQUESTED** [review]

File: Controller/Ajax/Analytics.php (hunk 5 above)
Comment (ahuininga-orisha) [reply Controller/Ajax/Analytics.php]: Implemented changes

File: Model/Client.php (hunk 6 above)
Comment (ahuininga-orisha) [reply Model/Client.php]: Implemented changes

File: view/frontend/templates/product/page.phtml
Comment (ahuininga-orisha) [reply view/frontend/templates/product/page.phtml:1]: Implemented changes

File: Model/Client.php [hunk 7]
```diff
@@ -134,12 +163,11 @@ protected function createHttpRequest(Request $tweakwiseRequest): HttpRequest
 
         if ($tweakwiseRequest->getParameters()) {
             $query = http_build_query($tweakwiseRequest->getParameters());
-            $url = sprintf('%s?%s', $url, $query);
+            $url   = sprintf('%s?%s', $url, $query);
```
Comment (jansentjeu) [inline Model/Client.php]: Space

**Review by jansentjeu — CHANGES_REQUESTED** [review]

File: Model/Client.php (hunk 7 above)
Comment (ahuininga-orisha) [reply Model/Client.php]: Removed it

**Review by jansentjeu — APPROVED** [review]

**Pull request merged**